### DISCUSSION TITLE
Using OpenCore Bootloader

### DISCUSSION THREAD

**Discussion by 0nelight** [0nelight/macOS-Disable-RAM-Areas#14] [General]
When using the Open Core Bootmanager (or Projects like [OpenCore Legacy Patcher](https://dortania.github.io/OpenCore-Legacy-Patcher/)) there is no need to use the EFI-Binary provided here.

Simply identify your faulty RAM-Area and change the config.plist File accordingly:

```batch
sudo diskutil mount disk0s1
patch /Volumes/EFI/EFI/OC/config.plist < patch.diff
```
```diff
--- config.plist
+++ config.plist
@@ -1362,7 +1362,20 @@
 			<false/>
 		</dict>
 		<key>ReservedMemory</key>
-		<array/>
+		<array>
+			<dict>
+				<key>Address</key>
+				<integer>12884901888</integer>
+				<key>Comment</key>
+				<string>Disable RAM-Area</string>
+				<key>Enabled</key>
+				<true/>
+				<key>Size</key>
+				<integer>167772160</integer>
+				<key>Type</key>
+				<string>UnusableMemory</string>
+			</dict>
+		</array>
 	</dict>
 </dict>
 </plist>
```

The _Address_ (12884901888 here) is the Start-Address and _Size_ (167772160 here) is the Area to disable. 
Please keep in mind that these two values **must be Page-Aligned**.

For details see https://dortania.github.io/docs/latest/Configuration.html#reservedmemory-properties.

**Update:**
The two Values must be provided in decimal-notation! Please go [here to check out](https://github.com/0nelight/macOS-Disable-RAM-Areas/discussions/14#discussioncomment-12291362) on how to calculate these two Values.

**Comment by bweh** [comment]
Hello there, I am using opencore legacy patcher on a device that has bad ram. i was about to use the efi file using refind until i found this thread.

I ran memtest86 and i got the following:
Lowest Error Address: 0x20C06B5E8 (8791700968)
Highest Error Address: 0x20E06B908 (8825256200)

So I edited my config.plist as you have shown above, with the Address being **8791700968** and the Size being **33558528**. However, when I shut the computer down and reboot, I initially got a blank screen with the Apple logo screen that just stalled. and then I tried resetting NVRAM which now leads me to this screen every time i boot.

![image](https://github.com/user-attachments/assets/736fbee2-66bd-47ed-bc21-6234d1c88415)

To be honest, I am unfamiliar with the process of determining the size and ensuring that it is page-aligned. I just subtracted the two addresses, divided by 4096, got the ceiling and then multiplied by 4096 to get the size.

Are you able to verify the process? Any help would be appreciated 🙏🏻

**Comment by 0nelight** [comment]
My AI-Slave wrote this Python script for you:

```python
def calculate_memory_reservation(A_low, A_high, page_size=4096):

    if isinstance(A_low, str):
        A_low = int(A_low, 0)
    elif not isinstance(A_low, int):
        raise ValueError("A_low must be a string or an integer")

    if isinstance(A_high, str):
        A_high = int(A_high, 0)
    elif not isinstance(A_high, int):
        raise ValueError("A_high must be a string or an integer")

    if A_low > A_high:
        raise ValueError("A_low must be less than or equal to A_high")

    # Calculate page-aligned start address
    page_low = A_low // page_size
    A_start = page_low * page_size

    # Calculate number of pages and total size
    page_high = A_high // page_size
    num_pages = page_high - page_low + 1
    S = num_pages * page_size

    return A_start, S


A_low_dec = 8791700968
A_high_dec = 8825256200
A_start_dec, S_dec = calculate_memory_reservation(A_low_dec, A_high_dec)
print(f"Page-aligned start address: {A_start_dec} (decimal), {hex(A_start_dec)} (hexadecimal)")
print(f"Size of the memory region: {S_dec} bytes (decimal), {hex(S_dec)} (hexadecimal)")
``` 
You can run it here:
https://www.online-python.com/kuVh5NrTHs

The error in your calculation was that you also have to page-align the Start address.

Please tell me if this worked for you :-)

**Reply by bweh** [reply]
Thank you for the quick reply. So far, this has worked perfectly! I previously wasn't even able to even update to macOS Monterey and newer OSes (it would say error occurred during installation, for some reason), but it can handle any installation now using OCLP (I'm on a 2015 MBP). Thank you for your help and for your initiative to preserve the life of macbooks 🙏🏻

edit: I just tested [Rember](https://www.kelleycomputing.net/rember/), and I no longer get any errors 🙌🏻

**Comment by theapplecore** [comment]
This is amazing work. I’m trying to deal with the same thing right now and I am currently running Memtest86 on a 2019 MacBook Pro with 32GB that has memory issues. 

How exactly do you find the lowest and highest error address. 

Any help will be much appreciated!

**Reply by 0nelight** [reply]
I am not willing to answer questions for which there are plenty of possibilies for you to find the answer yourself. Read the readme, the discussions, use search engine or a LLM (in that order).

Me uploading this simple code snippet to fix my own ram issues, doesn't mean this is kind of a free full-service consulting offering for solving your ram problems.

**Reply by theapplecore** [reply]
I appreciate your response. My ignorance on this topic is regrettable. I was just looking for some friendly advice from someone who has achieved something that I am personally trying to do. 

I apologize. 

The test is still running its first pass. I’m going to dig a bit deeper and possibly return with a more educated and thought out approach. 

I really appreciate your time!

**Reply by theapplecore** [reply]
![IMG_2226](https://github.com/user-attachments/assets/71f74046-3630-4b72-b232-c65fd76b4ded)

This seems to be the trouble. The size isn’t adding up. 

I can calculate this into decimal form but it’s massive.

**Reply by 0nelight** [reply]
quite a big area where it found errors, but doesn't say anything about how big the areas of errors are. 

please use the "badram-pattern" for displaying the errors - then we can see how big they are. (provide screenshot).

read and understand this before continuing:
https://github.com/0nelight/macOS-Disable-RAM-Areas/discussions/17#discussioncomment-12336775

also keep in mind there are two memtest; memtest86 from passmark and memtest86+ (see the plus sign!) https://www.memtest.org/.

in all the readme and discussions here we refer to memtest86+ (I don't know how different the two programs are)

**Reply by theapplecore** [reply]
50% complete. Will post results. Thanks for your quick replies!

**Reply by theapplecore** [reply]
I tried for several days without success concerning the test. It would get to a little over an hour and freeze. 

Thank you again for your support and quick responses. If you have any other input regarding the software freezing, I'd boot that MacBook back up but for now it is back on the shelf 😞

**Reply by 0nelight** [reply]
you tried with passmark memtest86 or with memtest86+ ?

If you tried with one, try the other in "badram mode".
